### DISCUSSION TITLE
Why ever would this compile?

### DISCUSSION THREAD

**Discussion by KillTheMule** [typst/typst#4122] [General]
I've just had this weird thing where I have in my typst code:

```typst
 if counter(page).get().first > 1 
```

Note the missing brackets after `first`. When executed on CI, typst was able to compile this (no way to check the output though), but when running the app (which embeds typst as a library) locally it would dutifully fail with 

` main.typ: <Id source not found: 'file not found (searched at \main.typ)'>: 'cannot access fields on type array'`

It's a pretty long shot and I don't have much in the way of debugging but... anybody knows what's up with this? I'd certainly have preferred if this failed on CI, too.

**Comment by PgBiel** [comment]
Seems like there is a lot of missing context; please send more of the code, as well as how it is being tested on CI.

**Reply by KillTheMule** [reply]
Yeah, that's a bit problematic, I can't really share that code, sorry. Here's a bigger excerpt though:

```typst
#set page(
  paper: "a4",
  margin:(x: 1.5cm, y: 1cm),
  footer: [
    #line()
    #v(-0.7em)
    #align(center)[
      #text(size: 12pt, weight: "regular")[
        Seite #context counter(page).display( "1 von 1", both: true)
      ]
    ]
  ],
  footer-descent: 5%,
  header: context {
    if counter(page).get().first() > 1 [
      #align(center)[
        #text(size: 12pt, weight: "regular")[
          Sometexthere
        ]
      ]
      #v(-0.7em)
      #line()
    ]
  },
  header-ascent: 5%
)
```

The whole document is a string inside the binary, where I'm inserting data via `sys.inputs`. I have a test that simply calls the document-producing top-level function of my crate and checks for success (i.e. it's run via `cargo test`), and the binary (which is built on the same  CI run) also calls this exact function (when using the right buttons). Only in the built binary I encounter the above error. Running the tests locally did also not exhibit this error (I'm developing and testing on linux, CI runs on windows though).

Sorry, I realize this is somewhat handwavy, and I'm not confident anyone can debug this this way, but I figured I'd need to ask if I'm missing something obvious.

**Reply by PgBiel** [reply]
`if condition { [markup] }` is the same as `if condition [markup]` (the latter is syntactic sugar for the former). Your error is unrelated.

**Reply by KillTheMule** [reply]
Not sure I follow. The above code is correct and works fine. Removing the brackets after `first` is what triggered the odd behavior. Curly brackets not being present anywhere indeed does not factor in here.

**Reply by PgBiel** [reply]
If removing [ ] doesn't trigger an error for wrongful #, then that code is likely not being run at all 🤔

**Reply by KillTheMule** [reply]
Oh wait, maybe I was unclear. Removing `[]` is out of the question, it's the `()` that wasn't there, like `.first` instead of `.first()`.

**Reply by PgBiel** [reply]
Then the error is correct: arrays don't have the `.first` field. But they do have the `.first()` method.

But you'll only trigger the error in CI if that piece of code is effectively run.

**Reply by KillTheMule** [reply]
Yeah that's what I'm saying, the error is correct, but didn't trigger in a case where I'm very sure it's run. The test is calling the same function as the binary, after all.

Maybe it's just one of them heisenbugs ;)

**Reply by PgBiel** [reply]
Unfortunately I cannot reproduce locally, with the given code (it does error). Are you sure you're compiling the file with those exact contents in your CI? Also, have you tested using the Typst CLI in your machine? Perhaps the command fails in CI but the failure isn't being picked up.

**Reply by KillTheMule** [reply]
Ok, this took quite some digging, but it was indeed a borked test (it did not run through all documents after all...). Thanks for bearing with me!

**Reply by PgBiel** [reply]
Great! Nice to know it all worked out in the end :)
